Access to nodes via VPN.

thz opened the pull request:
**What this PR does / why we need it**:

While the openvpn addon currently provides access to services
and pods, it does not provide access to the user cluster 
workers' kubelet ports.

This PR adds another exposed network for the vpn:
This `NodeAccessNetwork` is a private network which is used
to make hosts/nodes from user-cluster accessible from the
apiserver via the openvpn connection. This is achieved by
associating an IP address from that network to each host/node
and doing destination network address translation (DNAT)
to map the virtual IP addresses to the real host IP addresses.

The `NodeAccessNetwork` is specified as a flag to controller-manager
and used as templating data for creation of openvpn client and server
deployments.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Control plane can now reach the nodes via VPN
```

thz committed:
Introduce NodeAccessNetwork via VPN.

The NodeAccessNetwork is a private network which is used
to make hosts/nodes from user-cluster accessible from the
apiserver via the openvpn connection. This is achieved by
associating an IP address from that network to each host/node
and doing destination network address translation (DNAT)
to map the virtual IP addresses to the real host IP addresses.

* Adds NodeAccessNetwork to cluster ctx (controller-manager)
* Adds NodeAccessNetwork to ClusterTemplateData
* Adds iroute to openvpn-client-configs configmap
* Adds --push and --route to openvpn-server deployment
* Adds --node-access-network flag to controller-manager
* Adds a sidecar to openvpn-client (user-cluster) for DNAT-rules
* Adds basic firewalling to openvpn-server
* Updates kubermatic/openvpn image (in /containers now.)

mrIncompetent commented on addons/openvpn/openvpn-client-dep.yaml:
```diff
@@ -36,6 +37,44 @@ spec:
+          # iptables DNAT rules from it. no iterative approach, 
+          # no watching, but switching rules is atomic.
+          # real solution is a controller watching for nodes and
+          # adding/removing specific DNAT rules.
```
Please remove the comment and rather create a Issue for it.
As this is going to be visible to customers

mrIncompetent commented on addons/openvpn/openvpn-client-dep.yaml:
```diff
@@ -36,6 +37,44 @@ spec:
           name: openvpn-client-certificates
           readOnly: true
+      - name: mocked-dnat-controller
+        image: thzpub/vnf-swak-diag:1.2.1
```
Can't we use the existing Openvpn container?
If it's just missing iptables, i would recommend to simply add it to the image

thz replied:
Yes, this is prepared already and will be replaced tomorrow.

mrIncompetent commented on api/pkg/resources/openvpn/configmap.go:
```diff
@@ -29,24 +24,40 @@ func ConfigMap(data *resources.TemplateData, existing *corev1.ConfigMap) (*corev
+		net.IP(serviceNet.Mask).String()))
+
+	// iroute for virtual node network
+	if len(data.Cluster.Spec.ClusterNetwork.VirtualNodeNetwork) > 0 {
```
Why optional?

I would prefer always setting in within the controller (`ensureClusterNetworkDefaults`)
A default should be passed in via flag on the controller-manager

thz replied:
My thoughts were that some customers might not like the fact of exposing their workers by a vpn in this way. Complexity of making it optional seemed reasonable.
But: having it optional is optional too... So we can change it.

thz commented:
@mrIncompetent 
Updated as discussed. `NodeAccessNetwork` now going in as a flag to controller-manager.
Templating for vpn-client and vpn-server adds it to the deployments.
PTAL

thz commented:
rebased and squashed.

mrIncompetent commented on addons/openvpn/vpn-container/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.7
```
I would recomment we create a folder `/containers` on the root of the repository.
As the whole addons directory gets copied into a container later on and then gets used by the addon-controller, this might have unwanted sideeffects

mrIncompetent commented on addons/openvpn/vpn-container/release.sh:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
I would recomment we create a folder /containers on the root of the repository.
As the whole addons directory gets copied into a container later on and then gets used by the addon-controller, this might have unwanted sideeffects

mrIncompetent commented on api/cmd/kubermatic-controller-manager/main.go:
```diff
@@ -83,6 +84,7 @@ func main() {
 	flag.IntVar(&runOp.workerCount, "worker-count", 4, "Number of workers which process the clusters in parallel.")
 	flag.StringVar(&runOp.overwriteRegistry, "overwrite-registry", "", "registry to use for all images")
 	flag.StringVar(&runOp.nodePortRange, "nodeport-range", "30000-32767", "NodePort range to use for new clusters. It must be within the NodePort range of the seed-cluster")
+	flag.StringVar(&runOp.nodeAccessNetwork, "node-access-network", "", "A network which allows direct access to nodes via VPN. Uses CIDR notation.")
```
Please set a default value in the deployment

thz replied:
Done in 3e333ff.

thz commented:
- Applied node-access-network default in all relevant places (server+client deployment and configmap)
- fixed DNAT generation
- added MASQ for seed->user direction
-  ...and rebased.

mrIncompetent commented on api/pkg/resources/openvpn/configmap.go:
```diff
@@ -29,24 +24,45 @@ func ConfigMap(data *resources.TemplateData, existing *corev1.ConfigMap) (*corev
+
+	// iroute for node access network (default: 10.254.0.0/16)
+	nodeAccessNetwork := &net.IPNet{
+		IP:   net.IP{10, 254, 0, 0},
```
This should be a flag as well?

thz replied:
That _is_ the node access network. It is one of the places where the default is applied.

mrIncompetent commented on api/pkg/resources/openvpn/configmap.go:
```diff
@@ -29,24 +24,45 @@ func ConfigMap(data *resources.TemplateData, existing *corev1.ConfigMap) (*corev
+	if len(data.NodeAccessNetwork) > 0 {
+		_, nodeAccessNetwork, err = net.ParseCIDR(data.NodeAccessNetwork)
+		if err != nil {
+			return nil, fmt.Errorf("error creating vpn configmap while parsing node access network: %v", err)
```
Change to:
```
return nil, fmt.Errorf("failed to parse node access network %s: %v", data.NodeAccessNetwork, err)
```

This has 2 advantages:
- We're not adding the context of the function (`vpn configmap`) - as the caller already knows that this actually failed as it got an error.
- We're adding the value (`NodeAccessNetwork`) which causes the failure

thz replied:
Agreed, fixed.

mrIncompetent commented on api/pkg/resources/openvpn/deployment.go:
```diff
@@ -67,21 +67,87 @@ func Deployment(data *resources.TemplateData, existing *appsv1.Deployment) (*app
+
+	// node access network route (default: 10.254.0.0/16)
+	nodeAccessNetwork := &net.IPNet{
+		IP:   net.IP{10, 254, 0, 0},
```
This should be a flag as well?

thz replied:
The node access network must have the same value in three places.
How about moving applying the default upwards in the call stack? Maximum would be to set the default after parsing the flag in controller-manager.

mrIncompetent replied:
sgtm, adding it to `TemplateData`?

thz replied:
It is already there. Just the defaulting was not in a central place. This moved now to flag parsing.

mrIncompetent commented on api/pkg/resources/openvpn/deployment.go:
```diff
@@ -67,21 +67,87 @@ func Deployment(data *resources.TemplateData, existing *appsv1.Deployment) (*app
+	if len(data.NodeAccessNetwork) > 0 {
+		_, nodeAccessNetwork, err = net.ParseCIDR(data.NodeAccessNetwork)
+		if err != nil {
+			return nil, fmt.Errorf("error creating vpn deployment while parsing node access network: %v", err)
```
Change to:
```
return nil, fmt.Errorf("failed to parse node access network %s: %v", data.NodeAccessNetwork, err)
```

This has 2 advantages:
- We're not adding the context of the function (`vpn deployment`) - as the caller already knows that this actually failed as it got an error.
- We're adding the value (`NodeAccessNetwork`) which causes the failure

thz replied:
Agreed, fixed.
